libp2p-swarm/src/connection_reuse.rs

```diff
@@ -0,0 +1,186 @@
+use futures::stream::Fuse as StreamFuse;
+use multiaddr::Multiaddr;
+use muxing::StreamMuxer;
+use smallvec::SmallVec;
+use std::io::Error as IoError;
+use transport::{Transport, ConnectionUpgrade, UpgradedNode};
+
+/// Allows reusing the same muxed connection multiple times.
+///
+/// Can be created from an `UpgradedNode` through the `From` trait.
+///
+/// Implements the `Transport` trait.
+#[derive(Clone)]
+pub struct ConnectionReuse<T, C>
+	where T: Transport,
+		  C: ConnectionUpgrade<T::RawConn>
+{
+	// Underlying transport and connection upgrade for when we need to dial or listen.
+	inner: UpgradedNode<T, C>,
+}
+
+impl<T, C> From<UpgradedNode<T, C>> for ConnectionReuse<T, C>
+	where T: Transport,
+		  C: ConnectionUpgrade<T::RawConn>
+{
+	#[inline]
+	fn from(node: UpgradedNode<T, C>) -> ConnectionReuse<T, C> {
+		ConnectionReuse { inner: node }
+	}
+}
+
+impl<T, C> Transport for ConnectionReuse<T, C>
+	where T: Transport + 'static, // TODO: 'static :(
+		  C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :(
+		  C: Clone,
+		  C::Output: StreamMuxer + Clone,
+		  C::NamesIter: Clone // TODO: not elegant
+{
+	type RawConn = <C::Output as StreamMuxer>::Substream;
+	type Listener = ConnectionReuseListener<
+		Box<
+			Stream<
+				Item = (C::Output, Multiaddr),
+				Error = IoError,
+			>,
+		>,
+		C::Output,
+	>;
+	type Dial = Box<Future<Item = Self::RawConn, Error = IoError>>;
+
+	fn listen_on(self, addr: Multiaddr) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+		let (listener, new_addr) = match self.inner.listen_on(addr.clone()) {
+			Ok((l, a)) => (l, a),
+			Err((inner, addr)) => {
+				return Err((ConnectionReuse { inner: inner }, addr));
+			}
+		};
+
+		let listener = ConnectionReuseListener {
+			listener: listener.fuse(),
+			connections: Vec::new(),
+		};
+
+		Ok((listener, new_addr))
+	}
+
+	fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)> {
+		let dial = match self.inner.dial(addr) {
+			Ok(l) => l,
+			Err((inner, addr)) => {
+				return Err((ConnectionReuse { inner: inner }, addr));
+			}
+		};
+
+		let future = dial.and_then(|dial| dial.outbound());
+		Ok(Box::new(future) as Box<_>)
+	}
+}
+
+/// Implementation of `Stream<Item = (impl AsyncRead + AsyncWrite, Multiaddr)` for the
+/// `ConnectionReuse` struct.
+pub struct ConnectionReuseListener<S, M>
+	where S: Stream<Item = (M, Multiaddr), Error = IoError>,
+		  M: StreamMuxer
+{
+	listener: StreamFuse<S>,
+	connections: Vec<(M, <M as StreamMuxer>::InboundSubstream, Multiaddr)>,
+}
+
+impl<S, M> Stream for ConnectionReuseListener<S, M>
+	where S: Stream<Item = (M, Multiaddr), Error = IoError>,
+		  M: StreamMuxer + Clone + 'static // TODO: 'static :(
+{
+	type Item = (M::Substream, Multiaddr);
+	type Error = IoError;
+
+	fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+		match self.listener.poll() {
+			Ok(Async::Ready(Some((upgrade, client_addr)))) => {
+				let next_incoming = upgrade.clone().inbound();
```

Comment:
Ok, I said I was convinced before but I've become unconvinced. What's the point of taking `self` and then `clone`ing? It forces inner mutability for actual multiplexers but for raw connections it basically just adds an extra step over an `&self`-taking method. We might as well have a wrapper type for raw connections that `clone`s internally and implements `StreamMuxer`, dropping the `Clone` requirement entirely.